### PR TITLE
bpo-41818: Make additions to the tty module

### PR DESCRIPTION
**Pull request by 8vasu**
This follows #23536. Also, see #23686, #23740.

Add `cfmakeraw()`, `cfmakeecho()`, and `cfmakecbreak()`; make `setraw()` and `setcbreak()` return saved termios attributes to avoid extra `tcgetattr()` calls.

Post #23686, #23546, #23740 goals:
1. add `test_winsize()` to "Lib/test/test_pty.py";
2. major revision of "Lib/pty.py", which has not happened since the beginning of the millennium.

Notes
---
1. List of all POSIX.1-2017 input mode flags - https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/termios.h.html#tag_13_72_03_02.
2. List of all POSIX.1-2017 local mode flags - https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/termios.h.html#tag_13_72_03_06.
3. Naming justification for the functions: https://man7.org/linux/man-pages/man3/cfmakeraw.3.html Note that `cfmakeraw()` is not POSIX compliant; it is available on the BSDs and in glibc. However, POSIX has functions with similar names; for example, consider `cfgetispeed()`. Please see https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/termios.h.html.
4. `test_winch()` will be resubmitted as a part of a later pull request.
5. #23686 replaces `class winsize`.
6. #23740 replaces `tty.login_tty()`.

Signed-off-by: Soumendra Ganguly <soumendraganguly@gmail.com>

https://bugs.python.org/issue41818

### CODE REVIEW

**Comment by 8vasu** [comment]
@asvetlov @ethanfurman this follows #23536.

File: Lib/tty.py [hunk 1]
```diff
@@ -15,22 +24,81 @@
...
-def setraw(fd, when=TCSAFLUSH):
-    """Put terminal into a raw mode."""
-    mode = tcgetattr(fd)
-    mode[IFLAG] = mode[IFLAG] & ~(BRKINT | ICRNL | INPCK | ISTRIP | IXON)
-    mode[OFLAG] = mode[OFLAG] & ~(OPOST)
-    mode[CFLAG] = mode[CFLAG] & ~(CSIZE | PARENB)
-    mode[CFLAG] = mode[CFLAG] | CS8
-    mode[LFLAG] = mode[LFLAG] & ~(ECHO | ICANON | IEXTEN | ISIG)
+try:
+    CHECK_WINSZ = TIOCGWINSZ
```
Comment (asvetlov) [inline Lib/tty.py]: Duplicate

File: Lib/tty.py [hunk 2]
```diff
@@ -15,22 +24,81 @@
...
+        a tty, then OSError is raised."""
+        s = pack("HHHH", 0, 0, 0, 0)
+        return ioctl(fd, TIOCGWINSZ, s)
+
+    def setwinsz(fd, winsz):
+        """Sets window size of tty.
+        If fd is not a descriptor of
+        a tty, then OSError is raised."""
+        ioctl(fd, TIOCSWINSZ, winsz)
+except NameError:
```
Comment (asvetlov) [inline Lib/tty.py]: Explicit check for `hasattr(termios, "TIOCSWINSZ")` or something like this is better than `except NameError`.

File: Lib/tty.py [hunk 3]
```diff
@@ -15,22 +24,81 @@
...
+    HAVE_WINSZ = False
+    getwinsz = None
```
Comment (asvetlov) [inline Lib/tty.py]: I think better to make a stub that raises `NotImplementedError` is better than setting the function name to `None`.
In std lib we never use this pattern.

Sometimes we don't declare unsupported function at all, but in this case we usually mark in docs: "Availability: Windows" or "Availability: Linux kernel >= 4.5 or glibc >= 2.27".

I think here is better to support stubs that always fail.

File: Lib/tty.py [hunk 4]
```diff
@@ -15,22 +24,81 @@
...
+    setwinsz = None
+
+def mode_echo(mode, echo=True):
```
Comment (asvetlov) [inline Lib/tty.py]: `mode_echo` is not the best name I guess.
Looking on the function body I can suggest renaming with `enable_echo()`.

`enable_echo(mode)` and `enable_echo(mode, False)`.
Maybe others have better idea?

File: Lib/tty.py [hunk 5]
```diff
@@ -15,22 +24,81 @@
...
+    """Set/unset ECHO."""
+    if echo:
+        mode[LFLAG] |= ECHO
+    else:
+        mode[LFLAG] &= ~ECHO
+
+def mode_raw(mode):
```
Comment (asvetlov) [inline Lib/tty.py]: `enable_raw()` maybe if we agree with `enable_echo()`?

File: Lib/tty.py [hunk 6]
```diff
@@ -15,22 +24,81 @@
...
-    mode[CC][VTIME] = 0
-    tcsetattr(fd, when, mode)
+    new = list(mode)
+    new[LFLAG] &= ~(ECHO | ICANON)
+    new[CC][VMIN] = 1
+    new[CC][VTIME] = 0
+    tcsetattr(fd, when, new)
+    return mode
+
+def login(fd):
```
Comment (asvetlov) [inline Lib/tty.py]: The naming again.
For me, the function does not *login* but *makes the calling process a session leader* as you correctly wrote in docstring.

Maybe `make_session_leader()` or `set_session_leader()` is better?

**Review by asvetlov — CHANGES_REQUESTED** [review]
I suggest esablish public function names first, update the `tty` module documentation, and make the final review.

@jonathanslenders I very respect your work on https://github.com/prompt-toolkit/python-prompt-toolkit 
Maybe you want to review this (and follow-up) PR? Maybe you can suggest something? I really like to hear voices from experts in PTY/TTY area.

File: Lib/tty.py (hunk 1 above)
Comment (8vasu) [reply Lib/tty.py:28]: It did feel like there had to be a better way of doing this. I did not like the `except NameError` either. Thank you so much for this suggestion.

File: Lib/tty.py (hunk 2 above)
Comment (8vasu) [reply Lib/tty.py]: It did feel like there had to be a better way of doing this. I did not like the `except NameError` either. Thank you so much for this suggestion.

File: Lib/tty.py (hunk 1 above)
Comment (8vasu) [reply Lib/tty.py]: As you suggested, I will use `hasattr()` for this instead.

File: Lib/tty.py (hunk 3 above)
Comment (8vasu) [reply Lib/tty.py]: I am working on this.

File: Lib/tty.py (hunk 5 above)
Comment (8vasu) [reply Lib/tty.py]: Yes, that was a temporary name. This is supposed to function like `cfmakeraw(3)`. While `enable_raw()` does sound better than `mode_raw()`, we could also change it to `tty.makeraw()` if you permit me to do so, and if there are no licensing issues.

File: Lib/tty.py (hunk 4 above)
Comment (8vasu) [reply Lib/tty.py]: `enable_echo()` does sound better than `mode_echo()`. If you agree with `makeraw()` instead of `mode_raw()`, then we could also change this to `makeecho()`; however, I think that `enable_echo()` sounds better than either. I am leaning toward `enable_raw()`, and the updated commit will use that name unless someone else suggests something better.

File: Lib/tty.py (hunk 6 above)
Comment (8vasu) [reply Lib/tty.py]: Perfect! I will use one of those, or a variant of one of those. I agree that `tty.login()` was poor choice on my part; I wanted it to sound like `login_tty(3)`.

**Comment by 8vasu** [comment]
@asvetlov Thank you very much for the reviews. As per your suggestion, the updated commit will include documentation, etc.

File: Lib/tty.py (hunk 2 above)
Comment (8vasu) [reply Lib/tty.py]: I have one concern. To be able to use `hasattr(termios, "TIOCSWINSZ")`, I think I need to `import termios`. Am I correct about this? Since we already have `from termios import *`, which cannot be removed because of backward compatibility, is it good practice to also have an `import termios` there just for that single purpose? I was thinking that we could use

```
try:
   from termios import TIOCGWINSZ, TIOCSWINSZ
   ...
except ImportError:
   ...
```

which seems to be an explicit check as well.

File: Lib/tty.py (hunk 4 above)
Comment (8vasu) [reply Lib/tty.py]: How about `mkecho()`? See next section.

File: Lib/tty.py (hunk 5 above)
Comment (8vasu) [reply Lib/tty.py]: How about `mkraw()` instead of `makeraw()`? Reasoning: `mkecho()` seems better than `makeecho()`.

File: Lib/tty.py (hunk 6 above)
Comment (8vasu) [reply Lib/tty.py]: Update: since I had written these a while ago, I had forgotten what this function exactly does. I did some research, and it turns out that this function is actually doing a bit more than making the calling process a session leader.

According to https://pubs.opengroup.org/onlinepubs/9699919799/functions/setsid.html, the `setsid()` call in our function is responsible for making the calling process a session leader; moreover

> Upon return the calling process shall be the session leader of this new session, shall be the process group leader of a new process group, and **shall have no controlling terminal**.

Our function is also making `fd` the stdin/stdout/stderr and **controlling tty** of the calling process, which is why it belongs in the tty module.

Nevertheless, as you said, `login()` sounds like a bad choice; I completely agree with that. Should I change it to `login_tty()`? This function IS actually a clone of `login_tty(3)`.

Additional note: on my Debian system, `man login_tty` says

> These  are BSD functions, present in glibc. They are not standardized in POSIX.

File: Lib/tty.py (hunk 4 above)
Comment (asvetlov) [reply Lib/tty.py]: Sounds good

File: Lib/tty.py (hunk 5 above)
Comment (asvetlov) [reply Lib/tty.py]: Ok

File: Lib/tty.py (hunk 6 above)
Comment (asvetlov) [reply Lib/tty.py]: Maybe we should add `pty.login_tty()` function instead which calls the native `login_tty(3)`?
It can do some platform-specific things which we don't want to learn and reimplement in Python code.

**Comment by jonathanslenders** [comment]
@asvetlov : Thanks!

A few thoughts:

1.  For consistency, don't we need an ``mkcbreak`` as well?

2. The setraw implementation in prompt_toolkit is slightly different. But I'm certainly not an expert on this.
- See: https://github.com/prompt-toolkit/python-prompt-toolkit/blob/master/prompt_toolkit/input/vt100.py#L204
- It's a context manager.
- It's clearing IXOFF as well (not sure if needed).
- We clear INLCR and IGNCR to avoid translation of carriage return to newlines on input. This patch only clears ICRNL.
- There are a few other flags we don't clear in prompt_toolkit, but probably should: BRKINT, INPCK, ISTRIP, OPOST, CSIZE, PARENB.

3. I'm not familiar with the ``login_tty``  functionality.

4. setwinsize and getwinsize look fine. I've never used the pixel values though, and I've never done a "set size".

Overall I think this looks good.

One thing slightly related, but out of scope for this PR, that often missed in Python is that sys.stdin/sys.stdout are not context variables. So it's impossible to write an asyncssh server that exposes a REPL where print/input is working properly. (I guess I should propose that on python-ideas or bpo, but no energy/time to drive the discussion now.)

**Comment by 8vasu** [comment]
@asvetlov  Thank you for your patience. I have made the requested changes; please review again. I restructured the files a little bit. As a result, your reviews are showing up as outdated; they are not. Here are some notes to make reviewing easier for you:

1. Instead of using `hasattr(termios, "TIOCGWINSZ")`, I used

```
try:
    from termios import TIOCGWINSZ
    ...
except ImportError:
    ...
```
I did this to avoid `import termios`, since we already have a `from termios import *` which cannot be removed because of backward compatibility reasons. I understand that `from termios import *` and `import termios` can coexist; it just felt a little strange doing that :D  If you think that using `import termios` + `hasattr()` is the better option, then let me know; I will make the changes.

2. Did I add the stubs for `getwinsize()` and `setwinsize()` properly?

3. I have renamed the functions. One of the comments that I left on one of your reviews was related to the naming of `login_tty()`; please let me know if that is acceptable.

4. `getwinsize()` and `setwinsize()` now handle the packing/unpacking of structs. As the result, their input/output are tuples of integers instead of bytestrings. That makes them more user-friendly; this helped me remove the pack("HHHH", ...) from test_pty.py and abstract out the important parts.

5. I have written some documentation. I welcome suggestions!

I too would be honored if this work was reviewed by @jonathanslenders.

**Comment by jonathanslenders** [comment]
Could we return a namedtuple from getwinsize?

File: Lib/tty.py [hunk 7]
```diff
@@ -15,22 +24,93 @@
...
+    os.dup2(fd, STDOUT_FILENO)
+    os.dup2(fd, STDERR_FILENO)
+    if fd > STDERR_FILENO:
+        os.close(fd)
+
+    # The tty becomes the controlling terminal.
+    try:
+        ioctl(STDOUT_FILENO, TIOCSCTTY)
+    except (NameError, OSError):
+        tmp_fd = os.open(os.ttyname(STDOUT_FILENO), os.O_RDWR)
```
Comment (jonathanslenders) [inline Lib/tty.py]: I'm not sure what this line is doing exactly? Maybe add a comment here?

File: Lib/tty.py [hunk 8]
```diff
@@ -15,22 +24,93 @@
...
+    new[LFLAG] &= ~(ECHO | ICANON)
+    new[CC][VMIN] = 1
+    new[CC][VTIME] = 0
+    tcsetattr(fd, when, new)
+    return mode
+
+def login_tty(fd):
+    """Makes the calling process a session leader; if fd is a
+    descriptor of a tty, then that tty becomes the stdin, stdout,
+    stderr, and controlling tty of the calling process. Closes fd."""
```
Comment (jonathanslenders) [inline Lib/tty.py]: This comment implies that fd can either be or not be a descriptor of a tty.
Not sure what would happen if it's not a tty descriptor.

File: Lib/tty.py (hunk 7 above)
Comment (8vasu) [reply Lib/tty.py]: Here is the source of `login_tty(3)`: https://sourceware.org/git/?p=glibc.git;a=blob;f=login/login_tty.c

The
```
    except (NameError, OSError):
        tmp_fd = os.open(os.ttyname(STDOUT_FILENO), os.O_RDWR)
```
in our code is equivalent to their

```
 {
        /* This might work.  */
        char *fdname = ttyname (fd);
        int newfd;
        if (fdname)
          {
            if (fd != 0)
              (void) close (0);
            if (fd != 1)
              (void) close (1);
            if (fd != 2)
              (void) close (2);
            newfd = open (fdname, O_RDWR);
            (void) close (newfd);
          }
 }
```

I contacted the glibc people a while ago and this is what they had to say

> Since this was added by Roland almost 25 years ago, it seems to be trying to
emulate the TIOCSCTTY for system derived from System V (as indicated by this
stackflow thread [1]).

> Back in the day glibc tries to enclose fallbacks to be portable to the various
Unix implementations.  Currently this is just a dead code with various
shortcomings (as for a lot of the glibc fallback implementation where proper
kernel support was the correct answer) since neither Linux nor Hurd uses it.

> And it seems that Unixes derived from System V also has moved on from this
interface, neither Solaris nor Illumos seems to provide it.

> https://urldefense.com/v3/__https://stackoverflow.com/questions/51593530/code-explanation-for-glibc-login-tty-function-openttyname-immediately-f__;!!KwNVnqRv!Vtu1n8f09GnQ28wafOTRPvyTWxkgO0bd7c3sHMxR5GQqPjq4jJhaL_8J8kw-ZliJ$

TL;DR: `TIOCSCTTY` is available everywhere now. Therefore, I too think that it would be safe to remove the `except (NameError, OSError):` block. **Should I remove it?**

I am going to use `tty.login_tty()` [ or whatever we end up naming it ] to replace the fallback code in `pty.fork()`, which still uses the old System V approach. `tty.login_tty()` will also be used in `pty.spawn()` to make it easier for us to implement some new/improved functionality.

File: Lib/tty.py (hunk 8 above)
Comment (8vasu) [reply Lib/tty.py]: You are right. I will have to rephrase that. Actually, it will fail if `fd` is not a tty. **Do you think it would be a good idea to instead mention/document what exception would be raised if fd was not a tty?**

**Comment by 8vasu** [comment]
@asvetlov @jonathanslenders 

1. Yes! Named tuples sound like a good idea. Working on it!
2. `mkcbreak` will be added.
3.  Thank you for pointing us toward the `setraw()` implementation in prompt_toolkit. I will look into it after taking a nap :D My original idea was to make a clone of `cfmakeraw(3)`: https://sourceware.org/git/?p=glibc.git;a=blob;f=termios/cfmakeraw.c, which does what you mentioned except for clearing `INPCK`. **Should I make it clear all the flags that glibc's `cfmakeraw()` clears + maybe INPCK as well?**

**Comment by 8vasu** [comment]
@asvetlov @jonathanslenders I have made the requested changes; please review again. Notes follow.

1. Updated docs.
2. `mkcbreak()` implemented.
3. Our current mkraw now sets/unsets all flags mentioned by @jonathanslenders except that `IXOFF` is not being cleared. I think that it would not hurt to clear `IXOFF`; it is just that `cfmakeraw(3)` does not do it. **Should I go ahead and also clear `IXOFF`?** Since `IXOFF` is POSIX compliant, it will be present on every system that is immediately relevant.
4. `getwinsize()` now returns a namedtuple; the field names are same as that of the traditional `struct winsize`. The documentation is also advertising the `winsz` argument of `setwinsize()` as being a namedtuple of the same type.
5. A comment has been added to the `except (NameError, OSError):` clause in `tty.login_tty()`.
6. Docstring of `tty.login_tty()` has been modified to reflect the fact that `fd` must be a file descriptor of a tty.

**Comment by 8vasu** [comment]
https://pubs.opengroup.org/onlinepubs/7908799/xcurses/intov.html#tag_001_005_002 mentions that cbreak Mode must additionally have `ICRNL` cleared. The document is about "xcurses"; however, I checked using `stty` that even the python `curses.cbreak()` clears `ICRNL`. I am going to make another updated commit with very detailed comments!

**Comment by 8vasu** [comment]
I compared the following definitions/implementations of "raw mode tty" and "cbreak mode tty":

raw
-----
1. BSD `cfmakeraw()`
2. Python `curses.raw()`
3. `stty raw` on POSIX, Linux, BSD, and Solaris [ from manpages ]. The one on my Debian Linux system is the GNU coreutils `stty(1)`, which is the most general: it clears all the input flags that are there on the POSIX manpage of `stty`, which includes `IXOFF`.
4. prompt_toolkit's setraw!
5. the current python `tty.setraw()`

cbreak
---------
1. Python `curses.cbreak()`; based on what this does, and https://pubs.opengroup.org/onlinepubs/7908799/xcurses/intov.html#tag_001_005_002, our `mkcbreak()` function should also clear `ICRNL`.
2. `stty cbreak`
3. the current Python `tty.setcbreak()`

I would like to put down what I learned about `IXOFF` today: `IXON` enables Xon/Xoff "software flow control" on output, and `IXOFF` controls that on the input. Therefore, enabling or disabling `IXOFF` has no effect on `xterm(1)` or the ttys provided by the linux kernel (/dev/tty1, /dev/tty2,...); it is probably more relevant when working with a hardware termina; with a terminal emulator on a fast pc, that does not matter. See https://tldp.org/HOWTO/Text-Terminal-HOWTO-11.html, https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap11.html#tag_11_01_09. Both `C-s` and `C-q` are only affected when `IXON` are turned on or off; if `IXANY` is also enabled, then any character can be used to resume the flow control as opposed to just `C-q` (or whatever is the `START` character).

**Comment by 8vasu** [comment]
New commit: `winsize` is now a separate class to enforce nonnegative integer entries.

File: Doc/library/tty.rst [hunk 9]
```diff
@@ -3,35 +3,122 @@
...
 
-The :mod:`tty` module defines functions for putting the tty into cbreak and raw
-modes.
+The :mod:`tty` module defines various functions that perform common tty
+operations.
 
 Because it requires the :mod:`termios` module, it will work only on Unix.
 
-The :mod:`tty` module defines the following functions:
+Winsize Objects
```
Comment (8vasu) [inline Doc/library/tty.rst:21]: TODO: change "Winsize" to "winsize".

**Comment by 8vasu** [comment]
@jonathanslenders I have made the changes that you requested. I request you to please review this again when you have time.

Notes: `winsize` is now a class so that we can force the fields to be nonnegative. As you had hinted before, the `ws_xpixel` and `ws_ypixel` might be less useful now. I am not even sure if there exists an OS that still actually uses them. Should we remove them? They are certainly not hurting other functionality at the moment. The "get size" functionality already exists as `os.get_terminal_size()`: please see `os_get_terminal_size_impl()` in https://github.com/python/cpython/blob/master/Modules/posixmodule.c. However, that function also supports windows `CONIO`; further, there is no function for setting window size. I need both "get" and "set" window size functions (only on *nix) to be able to fix some problems with `pty`: see https://bugs.python.org/issue41818. Further, note that `shutil.get_terminal_size()` is not appropriate for this since it simply queries the variables `LINES` and `COLUMNS` instead of using `TIOCGWINSZ`. Thank you for you time.

**Comment by 8vasu** [comment]
`TIOCGWINSZ` and `TIOCSWINSZ` are widely available, but they are not POSIX compliant since POSIX generally does not standardize the `ioctl()` except for the STREAMS interface: see https://www.austingroupbugs.net/view.php?id=1151

They have decided to include functions `tcgetwinsize()` and `tcsetwinsize()` in IEEE Std 1003.1 ("POSIX.1") issue 8 [ the upcoming version of POSIX ]; see https://www.austingroupbugs.net/view.php?id=1151#c3856. NetBSD has already implemented them:

1. http://cvsweb.netbsd.org/bsdweb.cgi/src/lib/libc/termios/tcgetwinsize.c?only_with_tag=MAIN
2. http://cvsweb.netbsd.org/bsdweb.cgi/src/lib/libc/termios/tcsetwinsize.c?only_with_tag=MAIN

It should be noted that in the same issue, POSIX will also standardize `SIGWINCH`. It seems like they were originally going to name the functions `tcsetsize()`, `tcgetsize()`; but they have now renamed them to `tcsetwinsize()`, `tcgetwinsize()`. Please see:

1. https://www.opengroup.org/austin/docs/austin_839.txt
2. https://www.opengroup.org/austin/docs/austin_838.txt
3. https://www.opengroup.org/austin/docs/austin_833.txt
4. https://www.opengroup.org/austin/docs/austin_837.txt
5. https://www.opengroup.org/austin/docs/austin_834.txt
6. https://www.opengroup.org/austin/docs/austin_823.txt
7. https://www.opengroup.org/austin/docs/austin_836.txt
8. https://www.opengroup.org/austin/docs/austin_841.txt
9. https://www.opengroup.org/austin/docs/austin_913.txt
10. https://www.opengroup.org/austin/docs/austin_841.txt

Therefore, now I am highly tempted to change the `winsize` class back to a couple of functions named `tcsetwinsize()`, `tcgetwinsize()` OR at least, rename the methods to `tcsetwinsize()`, `tcgetwinsize()`. I need some help in making this decision. The benefit of having a class is that we can then have named attributes AND enforce nonnegative integer values for them.

**Comment by 8vasu** [comment]
@asvetlov @jonathanslenders Thank you for your patience. I have added the final test for `pty` to this pull request; please look for `test_winch()`. This test fails when current stdin is a tty and when `SIGWINCH` is available; it pinpoints the reason why `pty.spawn()` is unable to dynamically resize the slave window, which leads to gibberish output after, say, the current stdin is resized because of the creation of a tmux pane, or an emacs window in a frame, or when resizing the GUI [X/Wayland] terminal window. There should only be one more pull request after this :D

File: Doc/library/tty.rst [hunk 10]
```diff
@@ -1,41 +1,98 @@
...
+
+.. data:: HAVE_WINSZ
+
+   :const:`True` if both :const:`termios.TIOCGWINSZ` and :const:`termios.TIOCSWINSZ`
+   are defined; :const:`False` otherwise. The :class:`winsize` objects are implemented
+   if and only if this is :const:`True`.
+
+
+.. class:: winsize(row=0, col=0, xpixel=0, ypixel=0, fd=None)
+
```
Comment (8vasu) [inline Doc/library/tty.rst:32]: @jonathanslenders requested a namedtuple. I have made this class which serves the same purpose + ensures that the attributes are nonnegative integers; the traditional struct winsize has `unsigned short` members; I have done that by defining a __setattr__; please check tty.py.

File: Doc/library/tty.rst [hunk 11]
```diff
@@ -1,41 +1,98 @@
...
+   Create a :class:`winsize` object, which represents a struct winsize; the
+   attributes are *ws_row*, *ws_col*, *ws_xpixel*, and *ws_ypixel*; these are
+   all nonnegative integers. If *fd* is not :const:`None`, then get window
+   size of tty of which *fd* is a file descriptor. If *fd* is :const:`None`,
+   then set the attributes based on the values of *row*, *col*, *xpixel*, and
+   *ypixel*.
+
+
+.. method:: winsize.tcgetwinsize(fd)
+
```
Comment (8vasu) [inline Doc/library/tty.rst:42]: Naming justification: they have decided to include functions tcgetwinsize() and tcsetwinsize() in IEEE Std 1003.1 ("POSIX.1") issue 8 [ the upcoming version of POSIX ]; see https://www.austingroupbugs.net/view.php?id=1151#c3856.

File: Doc/library/tty.rst [hunk 12]
```diff
@@ -1,41 +1,98 @@
...
+   Set ECHO in the tty attribute list *mode*, which is a list like the one
+   returned by :func:`termios.tcgetattr`, if *echo* is :const:`True` or is
+   omitted. Unset ECHO if *echo* is :const:`False`.
+
+
+.. function:: cfmakeraw(mode)
+
+   Convert the tty attribute list *mode*, which is a list like the one returned
+   by :func:`termios.tcgetattr`, to that of a tty in raw mode.
+
```
Comment (8vasu) [inline Doc/library/tty.rst:71]: Naming justification: https://man7.org/linux/man-pages/man3/cfmakeraw.3.html

File: Lib/test/test_pty.py [hunk 13]
```diff
@@ -106,21 +98,23 @@ def setUp(self):
...
+                # Reset stdin window size as a part of the cleanup process
+                self.addCleanup(tty.winsize.tcsetwinsize, self.old_stdin_winsz,
+                                pty.STDIN_FILENO)
+            except tty.error:
                 pass
 
+        if tty.HAVE_WINCH:
+            old_winch = signal.getsignal(signal.SIGWINCH)
+            self.addCleanup(signal.signal, signal.SIGWINCH, old_winch)
+
```
Comment (8vasu) [inline Lib/test/test_pty.py:115]: Reset `SIGWINCH` handler.

File: Lib/test/test_pty.py [hunk 14]
```diff
@@ -142,47 +136,47 @@ def test_openpty(self):
...
+                debug("setting pty.STDIN_FILENO window size")
+                debug(f"original winsize: {self.old_stdin_winsz}")
+                target_stdin_winsz = tty.winsize()
+                target_stdin_winsz.ws_row = self.old_stdin_winsz.ws_row + 1
+                target_stdin_winsz.ws_col = self.old_stdin_winsz.ws_col + 1
+                target_stdin_winsz.ws_xpixel = self.old_stdin_winsz.ws_xpixel + 1
+                target_stdin_winsz.ws_ypixel = self.old_stdin_winsz.ws_ypixel + 1
+                debug(f"target winsize: {target_stdin_winsz}")
+                target_stdin_winsz.tcsetwinsize(pty.STDIN_FILENO)
+
```
Comment (8vasu) [inline Lib/test/test_pty.py:153]: These parts of the test were written by me. Now I am rewriting everything using the api that I am intending to add to tty.py; this increases code re-usability.

File: Lib/test/test_pty.py [hunk 15]
```diff
@@ -142,47 +136,47 @@ def test_openpty(self):
...
                                  "pty.STDIN_FILENO window size unchanged")
-            except OSError:
-                warnings.warn("Failed to set pty.STDIN_FILENO window size")
-                pass
+            except tty.error:
+                warnings.warn("failed to set pty.STDIN_FILENO window size")
 
         try:
-            debug("Calling pty.openpty()")
+            debug("calling pty.openpty()")
```
Comment (8vasu) [inline Lib/test/test_pty.py:163]: These are not necessarily complete sentences; therefore, they do not need to start with an uppercase letter/have a "." at the end. I feel that this looks better in the output. Thoughts?

File: Lib/test/test_pty.py [hunk 16]
```diff
@@ -324,6 +319,69 @@ def test_master_read(self):
         os.close(master_fd)
         self.assertEqual(data, b"")
 
+    @expectedFailureIfStdinIsTTYAndHAVE_WINCH
+    def test_winch(self):
```
Comment (8vasu) [inline Lib/test/test_pty.py:323]: New test! This demonstrates that `pty.spawn()` is currently incapable of dynamically resizing the slave end of the pty based on changes in size of the current stdin. See before and after pictures here: https://bugs.python.org/file49455/before.png, https://bugs.python.org/file49456/after.png

File: Lib/tty.py [hunk 17]
```diff
@@ -15,22 +23,136 @@
...
+    """Set/unset ECHO."""
+    if echo:
+        mode[LFLAG] |= ECHO
+    else:
+        mode[LFLAG] &= ~ECHO
+
+def cfmakeraw(mode):
+    """raw mode termios"""
+    # Clear all POSIX.1-2017 input mode flags.
+    mode[IFLAG] &= ~(IGNBRK | BRKINT | IGNPAR | PARMRK | INPCK | ISTRIP |
```
Comment (8vasu) [inline Lib/tty.py:36]: List of all POSIX.1-2017 input mode flags - https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/termios.h.html#tag_13_72_03_02

File: Lib/tty.py [hunk 18]
```diff
@@ -15,22 +23,136 @@
...
+        def __str__(self):
+            return f"(row={self.ws_row}, col={self.ws_col}, xpixel={self.ws_xpixel}, ypixel={self.ws_ypixel})"
+
+        def tcgetwinsize(self, fd):
+            """Gets window size of tty of which fd is a file descriptor."""
+            s = pack("HHHH", 0, 0, 0, 0)
+            try:
+                w = unpack("HHHH", ioctl(fd, TIOCGWINSZ, s))
+            except OSError as e:
+                raise error(e.errno, os.strerror(e.errno)) from e # termios.error
```
Comment (8vasu) [inline Lib/tty.py:137]: The original author intended this to be an extension of the `termios` module. In fact, notice that there is a `from termios import *` at the beginning of the file; the author also uses `tty.error` instead of `termios.error` in "Lib/pty.py". Therefore, all functions/methods must raise `termios.error` [ which is the same as `tty.error` ] when an `ioctl()`/`tcgetattr()`/`tcsetattr()`/... on `fd` [ which is expected to be the fd of a tty ] fails. Anyway, as mentioned before, `tcgetwinsize()` and `tcsetwinsize()` declarations are set to appear in `<termios.h>` as a part of IEEE Std 1003.1 ("POSIX.1") issue 8 [ the upcoming version of POSIX ]; see https://www.austingroupbugs.net/view.php?id=1151#c3856. Therefore, when our `winsize.tcgetwinsize()` and `winsize.tcsetwinsize()` methods fail, `termios.error` must be raised. That was one of the reasons why I removed `login_tty()` from this file; it has got almost nothing to do with `termios`. I will include it in the next pull request as a part of a different file.

File: Lib/tty.py [hunk 19]
```diff
@@ -15,22 +23,136 @@
...
+
+    class winsize:
+
+        def __init__(self, row=0, col=0, xpixel=0, ypixel=0):
+            raise NotImplementedError("termios.TIOCGWINSZ and/or termios.TIOCSWINSZ undefined")
+
+if HAVE_WINSZ and hasattr(signal, "SIGWINCH"):
+    HAVE_WINCH = True
+else:
+    HAVE_WINCH = False
```
Comment (8vasu) [inline Lib/tty.py:158]: Do we have `SIGWINCH`? :)

File: Lib/tty.py (hunk 6 above)
Comment (8vasu) [reply Lib/tty.py]: Sir, this sounds like a great idea. I have removed `tty.login_tty()` from this pull request. I have also done restructuring/cleanup. Please check the comments on the diff.

File: Lib/tty.py (hunk 5 above)
Comment (8vasu) [reply Lib/tty.py]: I have changed the names to `cfmakeraw()`, `cfmakeecho()`, ... so that they conform to standard naming convention.

File: Lib/tty.py [hunk 20]
```diff
@@ -15,22 +23,136 @@
...
+                return False
+
+        def __repr__(self):
+            return f"winsize(ws_row={self.ws_row}, ws_col={self.ws_col}, ws_xpixel={self.ws_xpixel}, ws_ypixel={self.ws_ypixel})"
+
+        def __str__(self):
+            return f"(row={self.ws_row}, col={self.ws_col}, xpixel={self.ws_xpixel}, ypixel={self.ws_ypixel})"
+
+        def tcgetwinsize(self, fd):
+            """Gets window size of tty of which fd is a file descriptor."""
```
Comment (8vasu) [inline Lib/tty.py:132]: @asvetlov Should tcgetwinsize() and tcsetwinsize() be defined in C as a part of termios.c [ the termios module ]? When tcgetwinsize() and tcsetwinsize() become standardized as a part of POSIX.1 issue 8, the C functions can be updated to use the native versions in termios.h; currently, they can simply use ioctl()+TIOCGWINSZ/TIOCSWINSZ.
If yes, then it would not be a bad idea to write cfmakeraw(), cfmakeecho(), and cfmakecbreak() etc as a part of termios.c either. Then tty.setraw() and tty.setcbreak() can simply make calls to those functions from the termios module. The "tty" library would then simply exist for backward compatibility; the original author of pty had created the tty module just to include the two helper functions tty.setraw() and tty.setcbreak(); it serves no other purpose.

**Comment by 8vasu** [comment]
Work on this should be temporarily suspended; see #23686

File: Lib/tty.py (hunk 6 above)
Comment (8vasu) [reply Lib/tty.py]: Please see #23740 for my implementation of `os.login_tty()`.

**Comment by 8vasu** [comment]
@jonathanslenders Sir, can you please review this again? I have removed a lot of the code [ `winsize` and `login_tty()` ]. Thank you in advance.

**Comment by 8vasu** [comment]
Still waiting for change review.

**Comment by 8vasu** [comment]
@asvetlov @gpshead Please take a look at this if time permits. You can ignore all comments made before since most of that is now covered by `termios.tcgetwinsize()`, `termios.tcsetwinsize()`, and `os.login_tty()`.

New summary for this PR: add `cfmakeraw()`, `cfmakeecho()`, and `cfmakecbreak()`; make `setraw()` and `setcbreak()` return saved termios attributes to avoid extra `tcgetattr()` calls.
